pkg/statedb/api_handler.go

```diff
@@ -4,28 +4,72 @@
 package statedb
 
 import (
+	"encoding/base64"
+	"encoding/gob"
 	"net/http"
 
 	"github.com/go-openapi/runtime"
 	"github.com/go-openapi/runtime/middleware"
 
 	. "github.com/cilium/cilium/api/v1/server/restapi/statedb"
 	restapi "github.com/cilium/cilium/api/v1/server/restapi/statedb"
+	"github.com/cilium/cilium/pkg/api"
 )
 
 func newDumpHandler(db *DB) restapi.GetStatedbDumpHandler {
-	return &dumphandler{db}
+	return &dumpHandler{db}
 }
 
-// REST API handler for the StateDB, handles JSON Dumping. The dump is also
-// available through `cilium statedb dump`, and is included in sysdumps.
-type dumphandler struct {
+// REST API handler for the '/statedb/dump' to dump the contents of the database
+// as JSON. Available through `cilium statedb dump` and included in sysdumps.
+type dumpHandler struct {
 	db *DB
 }
 
-// /statedb/dump
-func (h *dumphandler) Handle(params GetStatedbDumpParams) middleware.Responder {
+func (h *dumpHandler) Handle(params GetStatedbDumpParams) middleware.Responder {
 	return middleware.ResponderFunc(func(w http.ResponseWriter, _ runtime.Producer) {
 		h.db.ReadTxn().WriteJSON(w)
 	})
 }
+
+// REST API handler for '/statedb/query' to perform remote Get() and LowerBound()
+// queries against the database from 'cilium-dbg'.
+func newQueryHandler(db *DB) restapi.GetStatedbQueryTableHandler {
+	return &queryHandler{db}
+}
+
+type queryHandler struct {
+	db *DB
+}
+
+// /statedb/query
+func (h *queryHandler) Handle(params GetStatedbQueryTableParams) middleware.Responder {
+	txn := h.db.ReadTxn()
+	indexTxn := txn.getTxn().indexReadTxn(params.Table, params.Index)
+
+	iter := indexTxn.Root().Iterator()
+
+	key, err := base64.StdEncoding.DecodeString(params.Key)
+	if err != nil {
+		return api.Error(400, err)
+	}
+
+	if params.Lowerbound {
+		iter.SeekLowerBound(key)
+	} else {
+		iter.SeekPrefixWatch(key)
+	}
+
+	return middleware.ResponderFunc(func(w http.ResponseWriter, _ runtime.Producer) {
+		enc := gob.NewEncoder(w)
+		for _, obj, ok := iter.Next(); ok; _, obj, ok = iter.Next() {
```

Comment:
Is there a context <-done case we should be handling here? i.e. if the client closes the connection and the iterator doesn't produce any more objects then could this leak a goroutine?

Reply:
If the client closes connection then the `w.Write` call done by the `gob.Encoder` will return an error and thus `enc.Encode` will return an error and we return from this function.